theano/compile/builders.py

```diff
@@ -97,19 +99,40 @@ def __init__(self, inputs, outputs, **kwargs):
 
         self.new_inputs = new_inputs
         self.new_outputs = new_outputs
+        self.nin = len(new_inputs)
+        self.nout = len(new_outputs)
         self.inputs = inputs
         self.outputs = outputs
         self.kwargs = kwargs
         self.input_types = [input.type for input in inputs]
         self.output_types = [output.type for output in outputs]
 
+        if not hasattr(self, "fn"):
+            self.fn = orig_function(self.new_inputs,
+                                    self.new_outputs,
+                                    **self.kwargs)
+            self.fgraph = gof.FunctionGraph(self.fn.maker.fgraph.inputs,
+                                            self.fn.maker.fgraph.outputs)
+
+        self.init_c_code()
+
     def __eq__(self, other):
-        # TODO: recognize a copy
-        return self is other
+        if self is other:
+            return True
+        if (type(self) != type(other) or
+                self.nin != other.nin or
+                self.nout != other.nout):
+            return False
+        # see __hash__ for comment on why there is no mention of fgraph
+        # or module cache key here.
+        return (self._c_code == other._c_code)
 
     def __hash__(self):
         # TODO: use internal variables in hash
-        return hash(type(self))
+        return  hash((type(self),
+                      self.nin,
+                      self.nout,
+                      self._c_code))
```

Comment:
This (and \__eq__) could be replaced by `__props__ = ('nin', 'nout', '_c_code')`